File: sources/scalac/transformer/matching/VariableTraverser.java
package scalac.transformer.matching ;


import scalac.ast.Tree;
import scalac.util.Name;
import scalac.symtab.Symbol ;
import scalac.ast.Traverser ;

import Tree.Ident;
import Tree.Bind;


abstract class VariableTraverser extends Traverser {

      boolean isVariableName( Name name ) {
            return ( name.isVariable() ) && ( name != Name.fromString("_") ) ;

      }

      boolean isVariableSymbol( Symbol sym ) {
            return ( sym != null )&&( !sym.isPrimaryConstructor() );
      }

      abstract void handleVariableSymbol( Symbol sym ) ;

      public VariableTraverser() {
            super();
      }


      public void traverse(Tree tree) {
            switch (tree) {
            case Ident(Name name):
                  Symbol sym;

                  if( isVariableName( name )
                      && isVariableSymbol( sym = tree.symbol() ) )
                        handleVariableSymbol( sym );

                  return;

            case Bind(Name name, Tree subtree):
                  Symbol sym;

                  if( isVariableName( name )
                      && isVariableSymbol( sym = tree.symbol() ))
                        handleVariableSymbol( sym );

                  traverse( subtree );

                  return;

            case Select(_,_):
                  return;
            default:
                  super.traverse( tree );
            }
      }


}

File: sources/scala/tools/scalac/transformer/matching/VariableTraverser.scala
import scalac.ast.Tree;
import scalac.util.Name;
import scalac.symtab.Symbol ;
import scalac.ast.Traverser ;

import Tree.Ident;
import Tree.Bind;
import Tree.Select;

package scala.tools.scalac.transformer.matching {

abstract class VariableTraverser extends Traverser {

  def isVariableName(name: Name): Boolean = {
    ( name.isVariable() ) && ( name != Name.fromString("_") ) ;
  }

  def isVariableSymbol(sym: Symbol): Boolean = {
    ( sym != null )&&( !sym.isPrimaryConstructor() );
  }

  def handleVariableSymbol(sym: Symbol): Unit;

  override def traverse(tree: Tree): Unit = {
    tree match {
      case Ident(name)=>
        var sym: Symbol  = _;

        if( isVariableName( name )
           && isVariableSymbol( {sym = tree.symbol(); tree.symbol()} ) )
          handleVariableSymbol( sym );

        return;

      case Bind(name, subtree) =>
        var sym: Symbol = _;

        if( isVariableName( name )
           && isVariableSymbol( {sym = tree.symbol(); tree.symbol()} ))
          handleVariableSymbol( sym );

        traverse( subtree );

        return;

      case Select(_,_) =>
        return;

      case _ =>
        super.traverse( tree );
    }
  }


}
}
